### PR TITLE
Update mio to 0.8

### LINKED ISSUE #4135: Updating to Mio v0.8
For Mio v0.8 I plan on removing the `TcpSocket` type. Followed (closely, hopefully) by a v1.

`TcpSocket` functionality has grown beyond what Mio should support as it's by no means limited to provided non-blocking I/O anymore. The [socket2](https://docs.rs/socket2) crate provides all functionality and will be my recommended replacement, but I'm biased as I maintain both crate. However @carllerche already indicated he wanted to keep the dependencies low, so adding socket isn't a solution. So, I'm opening issue to discuss to upgrade to Mio v0.8 with `TcpSocket`.

P.S. I don't there will be any other breaking changes to Mio v0.8.

/cc @carllerche

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #4135 

Unblocks #3082, #3312, #3542

### CODE REVIEW

**Comment by taiki-e** [comment]
As for the socket2 issue, I don't think it should be considered a blocker, but if one of the reviewers wants to make it a blocker, I'm fine with that.

File: tokio/src/net/tcp/socket.rs
Comment (Darksonn) [inline tokio/src/net/tcp/socket.rs:323]: There are several casts in this PR. Do we risk them truncating any data here?

File: tokio/src/net/tcp/socket.rs
Comment (Darksonn) [inline tokio/src/net/tcp/socket.rs:506]: It seems like the `Socket` type in socket2 implements `IntoRawFd`, which seems more appropriate here than using `AsRawFd` with `ManuallyDrop`.

File: tokio/Cargo.toml
Comment (Darksonn) [inline tokio/Cargo.toml:57]: I'm not too familiar with the features of socket2, but can we be more precise here?

File: tokio/src/net/tcp/socket.rs
Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs:323]: There are 5 casts:
- set_send_buffer_size,set_recv_buffer_size (cast u32 as usize): on tokio, [usize is always at least 32 bits](https://github.com/tokio-rs/tokio/blob/cf3206842c0d94ecdaaeb421a58b1c963b627c3d/tokio/src/lib.rs#L357-L368).
- send_buffer_size,recv_buffer_size (cast usize as u32): socket2 gets these values as  c_int ([always i32](https://grep.app/search?q=type%20c_int%20%3D&filter[repo][0]=rust-lang/libc)) and then cast them to usize ([1](https://github.com/rust-lang/socket2/blob/cddb67b8534a46a372beb9fd63c5ce8d1473cbda/src/socket.rs#L857-L858), [2](https://github.com/rust-lang/socket2/blob/cddb67b8534a46a372beb9fd63c5ce8d1473cbda/src/socket.rs#L928-L929)).
- listen (cast u32 as i32): It seems [using cast is wrong](https://github.com/tokio-rs/mio/blob/77bcab0fe57343113700cbb0af76de44ab1ffadf/src/sys/unix/tcp.rs#L37), I'll fix this.

File: tokio/Cargo.toml
Comment (taiki-e) [reply tokio/Cargo.toml:57]: socket2's `all` feature means enabling APIs that are not available in some OSes, not means enabling all feature flags.

https://github.com/rust-lang/socket2/blob/6601ed132b37d6e9d178b34918bfb0b236800232/Cargo.toml#L42-L43

> ```toml
> # Enable all API, even ones not available on all OSs.
> all = []
> ```

In our case, this feature is needed to use [reuse_port](https://docs.rs/socket2/0.4.2/socket2/struct.Socket.html#method.reuse_port)/[set_reuse_port](https://docs.rs/socket2/0.4.2/socket2/struct.Socket.html#method.set_reuse_port), which is not available in Solaris & Illumos.

File: tokio/src/net/tcp/socket.rs
Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs:324]: fixed in https://github.com/tokio-rs/tokio/pull/4270/commits/5326df1888392e3240d8142e05b17c94c1001350

File: tokio/src/net/tcp/socket.rs
Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs]: done in https://github.com/tokio-rs/tokio/pull/4270/commits/216baeb220fb85bcfabb36b2ff20d44c0c3eb957

**Review by Darksonn — APPROVED** [review]
The change itself seems good, though I'd like to discuss it this evening when the americans come online.

**Review by Noah-Kennedy — APPROVED** [review]
This looks good.

File: tokio/src/net/tcp/socket.rs
Comment (zh-jq) [inline tokio/src/net/tcp/socket.rs:530]: How about `impl From<socket2::Socket> for TcpSocket`?

Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs:530]: We don't want to make a crate with version 0.x a public dependency of a crate with version 1.x.

See also api-guidelines: https://rust-lang.github.io/api-guidelines/necessities.html#public-dependencies-of-a-stable-crate-are-stable-c-stable

Comment (taiki-e) [reply tokio/src/net/tcp/socket.rs:530]: Also, once this PR is merged, adding the options that socket2 has to the TcpSocket will be easy. So the cases where code like this example is needed should be greatly reduced in the near future.

**Comment by taiki-e** [comment]
I got two approvals and no one had raised any objections for over a month, so I'll merge this.

**Pull request merged**
